src/main/java/seedu/address/model/history/VersionedAddressBook.java

```diff
@@ -0,0 +1,65 @@
+package seedu.address.model.history;
+
+import java.util.ArrayList;
+import java.util.Objects;
+
+import seedu.address.model.AddressBook;
+
+/**
+ * Represents the different states of the Address Book for the user.
+ */
+public class VersionedAddressBook {
+    private final ArrayList<AddressBook> addressBookStateList;
+    private int currentStatePointer;
+
+    /**
+     * Constructs a state list for the address book and saves the initial copy.
+     */
+    public VersionedAddressBook(AddressBook addressBook) {
+        addressBookStateList = new ArrayList<>();
+        currentStatePointer = 0;
+        addressBookStateList.add(new AddressBook(addressBook));
+    }
+
+    /**
+     * Saves the whole address book state in the history.
+     */
+    public void commitAddressBook(AddressBook addressBook) {
+        currentStatePointer++;
+        addressBookStateList.add(currentStatePointer, new AddressBook(addressBook));
+    }
+
+    /**
+     * Reverses the given AddressBook to the previous state.
+     */
+    public void undoAddressBook(AddressBook addressBook) {
+        assert currentStatePointer >= 0 : "This is the earliest version that user can retrieve";
```

Comment:
maybe an exception is more appropriate here, since it can be caused by the user undoing when history is empty.